source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -561,16 +578,15 @@ FilterDataStatus Filter::onData(ProcessorState& state, Buffer::Instance& data, b
     // Need to first enqueue the data into the chunk queue before sending.
     auto req = setupBodyChunk(state, data, end_stream);
     state.enqueueStreamingChunk(data, end_stream);
-    sendBodyChunk(state, ProcessorState::CallbackState::StreamedBodyCallback, req);
 
-    // At this point we will continue, but with no data, because that will come later
-    if (end_stream) {
-      // But we need to stop iteration for the last chunk because it's our last chance to do stuff
+    if (end_stream || state.callbackState() == ProcessorState::CallbackState::HeadersCallback) {
```

Comment:
If the callback state is waiting for headers response, the filter state to be returned here can not be continue as that will cause no chance to mutate the header.